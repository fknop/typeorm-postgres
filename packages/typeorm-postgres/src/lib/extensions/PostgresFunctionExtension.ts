import {RdbmsSchemaBuilderHook} from 'typeorm'

// These imports are not ideal, if TypeORM decide to change the locations of these files, it will not work anymore
import {SqlInMemory} from 'typeorm/driver/SqlInMemory'
import {Query} from 'typeorm/driver/Query'
import {RdbmsSchemaBuilder} from 'typeorm/schema-builder/RdbmsSchemaBuilder'
import {PostgresQueryRunner} from 'typeorm/driver/postgres/PostgresQueryRunner'

import {getPostgresMetadataArgsStorage} from '../metadata/TypeormPostgresArgsMetadata'
import {PostgresFunctionOptions} from '../options/PostgresFunctionOptions'

const METADATA_TYPE = 'POSTGRES_FUNCTION'

export class PostgresFunctionExtension {
  static register({
    functions,
  }: {
    functions: Function[]
  }): RdbmsSchemaBuilderHook {
    class PostgresExtension implements RdbmsSchemaBuilderHook {
      private existingFunctions: WeakMap<
        RdbmsSchemaBuilder,
        PostgresFunctionOptions<string>[]
      > = new WeakMap()

      private metadataTable = 'typeorm_metadata'

      async init(
        queryRunner: PostgresQueryRunner,
        schemaBuilder: RdbmsSchemaBuilder
      ): Promise<void> {
        this.metadataTable = queryRunner.getTypeormMetadataTableName()

        this.existingFunctions.set(
          schemaBuilder,
          await this.getCurrentFunctions(queryRunner)
        )
      }

      beforeAll(
        queryRunner: PostgresQueryRunner,
        schemaBuilder: RdbmsSchemaBuilder
      ): Promise<SqlInMemory> {
        const sqlInMemory = new SqlInMemory()

        const args = getPostgresMetadataArgsStorage()
        const functionsToSync: PostgresFunctionOptions<string>[] = []

        functions?.forEach((func) => {
          const functionToSync = args.filterFunctions(func)

          if (functionToSync) {
            functionsToSync.push(functionToSync.options)
          }
        })

        const existingFunctions =
          this.existingFunctions.get(schemaBuilder) ?? []

        const functionsQueries = this.getFunctionsQueries(queryRunner, {
          existingFunctions,
          functionsToSync,
        })

        sqlInMemory.upQueries.push(...functionsQueries.upQueries)
        sqlInMemory.downQueries.push(...functionsQueries.downQueries)

        return Promise.resolve(sqlInMemory)
      }

      private async getCurrentFunctions(
        queryRunner: PostgresQueryRunner
      ): Promise<PostgresFunctionOptions<string>[]> {
        const hasTable = await queryRunner.hasTable(this.metadataTable)

        if (!hasTable) {
          return []
        }

        const qb = await queryRunner.connection.createQueryBuilder()

        const metadatas: {name: string; value: string}[] = await qb
          .select()
          .from(this.metadataTable, 't')
          .where(`type = :type`, {
            type: METADATA_TYPE,
          })
          .getRawMany()

        return metadatas.map((metadata) => {
          // Might need to handle this a little better
          return {
            name: metadata.name,
            definition: metadata.value,
          } as PostgresFunctionOptions<string>
        })
      }

      private getFunctionsQueries(
        queryRunner: PostgresQueryRunner,
        {
          existingFunctions,
          functionsToSync,
        }: {
          existingFunctions: PostgresFunctionOptions<string>[]
          functionsToSync: PostgresFunctionOptions<string>[]
        }
      ): SqlInMemory {
        const sqlInMemory = new SqlInMemory()

        const functionsToDrop: PostgresFunctionOptions<string>[] = []
        const functionToCreate: PostgresFunctionOptions<string>[] = []
        const functionsToUpdate: {
          old: PostgresFunctionOptions<string>
          new: PostgresFunctionOptions<string>
        }[] = []

        existingFunctions.forEach((existingFunction) => {
          const functionToSync = functionsToSync.find(
            (functionToSync) => existingFunction.name === functionToSync.name
          )

          if (!functionToSync) {
            functionsToDrop.push(existingFunction)
          } else if (this.isUpdatedFunction(existingFunction, functionToSync)) {
            functionsToUpdate.push({
              old: existingFunction,
              new: functionToSync,
            })
          }
        })

        functionsToSync.forEach((functionToSync) => {
          if (
            !existingFunctions.find((func) => func.name === functionToSync.name)
          ) {
            functionToCreate.push(functionToSync)
          }
        })

        functionsToDrop.forEach((functionToDrop) => {
          sqlInMemory.upQueries.push(this.getDropFunctionQuery(functionToDrop))
          sqlInMemory.downQueries.push(
            this.getCreateFunctionQuery(functionToDrop)
          )

          sqlInMemory.upQueries.push(
            queryRunner.deleteTypeormMetadataSql({
              // Typeorm doesn't allow any string in the type parameter - should we reimplement this method?
              type: METADATA_TYPE as any,
              name: functionToDrop.name,
            })
          )
          sqlInMemory.downQueries.push(
            queryRunner.insertTypeormMetadataSql({
              name: functionToDrop.name,
              // Typeorm doesn't allow any string in the type parameter - should we reimplement this method?
              type: METADATA_TYPE as any,
              value: functionToDrop.definition,
            })
          )
        })

        functionToCreate.forEach((functionToCreate) => {
          sqlInMemory.upQueries.push(
            this.getCreateFunctionQuery(functionToCreate)
          )
          sqlInMemory.downQueries.push(
            this.getDropFunctionQuery(functionToCreate)
          )

          sqlInMemory.upQueries.push(
            queryRunner.insertTypeormMetadataSql({
              name: functionToCreate.name,
              // Typeorm doesn't allow any string in the type parameter - should we reimplement this method?
              type: METADATA_TYPE as any,
              value: functionToCreate.definition,
            })
          )
          sqlInMemory.downQueries.push(
            queryRunner.deleteTypeormMetadataSql({
              // Typeorm doesn't allow any string in the type parameter - should we reimplement this method?
              type: METADATA_TYPE as any,
              name: functionToCreate.name,
            })
          )
        })

        functionsToUpdate.forEach((func) => {
          sqlInMemory.upQueries.push(this.getCreateFunctionQuery(func.new))
          sqlInMemory.downQueries.push(this.getCreateFunctionQuery(func.old))
        })

        return sqlInMemory
      }

      private getCreateFunctionQuery(func: PostgresFunctionOptions<string>) {
        return new Query(func.definition)
      }

      private getDropFunctionQuery(func: PostgresFunctionOptions<string>) {
        return new Query(`DROP FUNCTION IF EXISTS "${func.name}"`)
      }

      private isUpdatedFunction(
        a: PostgresFunctionOptions<string>,
        b: PostgresFunctionOptions<string>
      ) {
        return a.definition !== b.definition
      }
    }

    return new PostgresExtension()
  }
}
