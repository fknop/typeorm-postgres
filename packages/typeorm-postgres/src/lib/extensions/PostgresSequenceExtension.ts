import {RdbmsSchemaBuilderHook} from 'typeorm'

// These imports are not ideal, if TypeORM decide to change the locations of these files, it will not work anymore
import {SqlInMemory} from 'typeorm/driver/SqlInMemory'
import {Query} from 'typeorm/driver/Query'
import {RdbmsSchemaBuilder} from 'typeorm/schema-builder/RdbmsSchemaBuilder'
import {PostgresQueryRunner} from 'typeorm/driver/postgres/PostgresQueryRunner'

import {getPostgresMetadataArgsStorage} from '../metadata/TypeormPostgresArgsMetadata'
import {PostgresSequenceOptions} from '../options/PostgresSequenceOptions'
import {isNil} from '../utils/isNil'

interface DatabaseSequenceRow {
  sequence_name: string
}

const METADATA_TYPE = 'POSTGRES_SEQUENCE'

export class PostgresSequenceExtension {
  static register({
    sequences,
  }: {
    sequences: Function[]
  }): RdbmsSchemaBuilderHook {
    class PostgresExtension implements RdbmsSchemaBuilderHook {
      private existingSequences: WeakMap<
        RdbmsSchemaBuilder,
        PostgresSequenceOptions[]
      > = new WeakMap()

      private metadataTable = 'typeorm_metadata'

      async init(
        queryRunner: PostgresQueryRunner,
        schemaBuilder: RdbmsSchemaBuilder
      ): Promise<void> {
        this.metadataTable = queryRunner.getTypeormMetadataTableName()

        this.existingSequences.set(
          schemaBuilder,
          await this.getCurrentSequences(queryRunner)
        )
      }

      beforeAll(
        queryRunner: PostgresQueryRunner,
        schemaBuilder: RdbmsSchemaBuilder
      ): Promise<SqlInMemory> {
        const sqlInMemory = new SqlInMemory()

        const args = getPostgresMetadataArgsStorage()
        const sequencesToSync: PostgresSequenceOptions[] = []

        sequences?.forEach((sequence) => {
          const sequenceToSync = args.filterSequences(sequence)

          if (sequenceToSync) {
            sequencesToSync.push(sequenceToSync.options)
          }
        })

        const existingSequences =
          this.existingSequences.get(schemaBuilder) ?? []
        const sequenceQueries = this.getSequencesQueries(queryRunner, {
          existingSequences,
          sequencesToSync,
        })

        sqlInMemory.upQueries.push(...sequenceQueries.upQueries)
        sqlInMemory.downQueries.push(...sequenceQueries.downQueries)

        return Promise.resolve(sqlInMemory)
      }

      /**
       * We're getting the current sequences inside the databases
       * We cannot simply rely on the "information_schema"."sequences" table because some sequences might not be managed by this extension
       * We use the typeorm metadata table to store whether we're managing a sequence
       */
      private async getCurrentSequences(
        queryRunner: PostgresQueryRunner
      ): Promise<PostgresSequenceOptions[]> {
        const allSequences: DatabaseSequenceRow[] = await queryRunner.query(
          `SELECT * FROM "information_schema"."sequences"`
        )

        const hasTable = await queryRunner.hasTable(this.metadataTable)

        if (!hasTable) {
          return []
        }

        const qb = await queryRunner.connection.createQueryBuilder()

        const metadatas: {name: string; value: string}[] = await qb
          .select()
          .from(this.metadataTable, 't')
          .where(`type = :type`, {
            type: METADATA_TYPE,
          })
          .getRawMany()

        const databaseSequences = allSequences.map(
          (metadata) => metadata.sequence_name
        )

        return metadatas
          .map((metadata) => {
            // Might need to handle this a little better
            return JSON.parse(metadata.value) as PostgresSequenceOptions
          })
          .filter((sequence) => databaseSequences.includes(sequence.name))
      }

      // Keep track of sequences managed by the extension
      private getSequencesQueries(
        queryRunner: PostgresQueryRunner,
        {
          existingSequences,
          sequencesToSync,
        }: {
          existingSequences: PostgresSequenceOptions[]
          sequencesToSync: PostgresSequenceOptions[]
        }
      ): SqlInMemory {
        const sqlInMemory = new SqlInMemory()

        const sequencesToDrop: PostgresSequenceOptions[] = []
        const sequencesToCreate: PostgresSequenceOptions[] = []
        const sequencesToUpdate: {
          old: PostgresSequenceOptions
          new: PostgresSequenceOptions
        }[] = []

        existingSequences.forEach((sequence) => {
          const sequenceToSync = sequencesToSync.find(
            (sequenceToSync) => sequence.name === sequenceToSync.name
          )
          if (!sequenceToSync) {
            sequencesToDrop.push(sequence)
          } else if (this.isUpdatedSequence(sequence, sequenceToSync)) {
            sequencesToUpdate.push({
              old: sequence,
              new: sequenceToSync,
            })
          }
        })

        sequencesToSync.forEach((sequenceToSync) => {
          if (
            !existingSequences.find(
              (sequence) => sequence.name === sequenceToSync.name
            )
          ) {
            sequencesToCreate.push(sequenceToSync)
          }
        })

        sequencesToDrop.forEach((sequence) => {
          sqlInMemory.upQueries.push(this.getDropSequenceQuery(sequence))
          sqlInMemory.downQueries.push(this.getCreateSequenceQuery(sequence))

          sqlInMemory.upQueries.push(
            queryRunner.deleteTypeormMetadataSql({
              // Typeorm doesn't allow any string in the type parameter - should we reimplement this method?
              type: METADATA_TYPE as any,
              name: sequence.name,
            })
          )
          sqlInMemory.downQueries.push(
            queryRunner.insertTypeormMetadataSql({
              name: sequence.name,
              // Typeorm doesn't allow any string in the type parameter - should we reimplement this method?
              type: METADATA_TYPE as any,
              value: JSON.stringify(sequence),
            })
          )
        })

        sequencesToCreate.forEach((sequence) => {
          sqlInMemory.upQueries.push(this.getCreateSequenceQuery(sequence))
          sqlInMemory.downQueries.push(this.getDropSequenceQuery(sequence))

          sqlInMemory.upQueries.push(
            queryRunner.insertTypeormMetadataSql({
              name: sequence.name,
              // Typeorm doesn't allow any string in the type parameter - should we reimplement this method?
              type: METADATA_TYPE as any,
              value: JSON.stringify(sequence),
            })
          )
          sqlInMemory.downQueries.push(
            queryRunner.deleteTypeormMetadataSql({
              // Typeorm doesn't allow any string in the type parameter - should we reimplement this method?
              type: METADATA_TYPE as any,
              name: sequence.name,
            })
          )
        })

        sequencesToUpdate.forEach((sequence) => {
          sqlInMemory.upQueries.push(this.getAlterSequenceQuery(sequence.new))
          sqlInMemory.downQueries.push(this.getAlterSequenceQuery(sequence.old))
        })

        return sqlInMemory
      }

      private getCreateSequenceQuery(sequence: PostgresSequenceOptions) {
        return new Query(
          `CREATE SEQUENCE "${sequence.name}"${this.getSequenceParameters(
            sequence
          )}`
        )
      }

      private getAlterSequenceQuery(sequence: PostgresSequenceOptions) {
        return new Query(
          `ALTER SEQUENCE "${sequence.name}"${this.getSequenceParameters(
            sequence
          )}`
        )
      }

      private getSequenceParameters(sequence: PostgresSequenceOptions) {
        return ` AS ${sequence.dataType}${sequence.cycle ? ' CYCLE' : ''}${
          !isNil(sequence.start) ? ` START ${sequence.start}` : ''
        }${!isNil(sequence.minValue) ? ` MINVALUE ${sequence.minValue}` : ''}${
          !isNil(sequence.maxValue) ? ` MAXVALUE ${sequence.maxValue}` : ''
        }${sequence.increment ? ` INCREMENT ${sequence.increment}` : ''}`
      }

      private getDropSequenceQuery(sequence: PostgresSequenceOptions) {
        return new Query(`DROP SEQUENCE IF EXISTS "${sequence.name}"`)
      }

      private isUpdatedSequence(
        a: PostgresSequenceOptions,
        b: PostgresSequenceOptions
      ) {
        return (
          a.cycle !== b.cycle ||
          a.dataType !== b.dataType ||
          // Using double equal instead of triple because of number/bigint comparisons
          a.start != b.start ||
          a.minValue != b.minValue ||
          a.maxValue != b.maxValue ||
          a.increment != b.increment
        )
      }
    }

    return new PostgresExtension()
  }
}
