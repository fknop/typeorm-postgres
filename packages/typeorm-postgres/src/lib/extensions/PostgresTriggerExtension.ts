import {EntityMetadata, RdbmsSchemaBuilderHook} from 'typeorm'

// These imports are not ideal, if TypeORM decide to change the locations of these files, it will not work anymore
import {SqlInMemory} from 'typeorm/driver/SqlInMemory'
import {Query} from 'typeorm/driver/Query'
import {RdbmsSchemaBuilder} from 'typeorm/schema-builder/RdbmsSchemaBuilder'
import {PostgresQueryRunner} from 'typeorm/driver/postgres/PostgresQueryRunner'

import {getPostgresMetadataArgsStorage} from '../metadata/TypeormPostgresArgsMetadata'
import {PostgresTriggerOptions} from '../options/PostgresTriggerOptions'

const METADATA_TYPE = 'POSTGRES_TRIGGER'

type PostgresTrigger = PostgresTriggerOptions & {
  tableName: string
  functionName: string
}

export class PostgresTriggerExtension {
  static init(): RdbmsSchemaBuilderHook {
    class PostgresExtension implements RdbmsSchemaBuilderHook {
      private existingTriggers: WeakMap<
        RdbmsSchemaBuilder,
        {name: string; definition: string}[]
      > = new WeakMap()

      private metadataTable = 'typeorm_metadata'

      async init(
        queryRunner: PostgresQueryRunner,
        schemaBuilder: RdbmsSchemaBuilder
      ): Promise<void> {
        this.metadataTable = queryRunner.getTypeormMetadataTableName()

        this.existingTriggers.set(
          schemaBuilder,
          await this.getCurrentTriggers(queryRunner)
        )
      }

      afterAll(
        queryRunner: PostgresQueryRunner,
        schemaBuilder: RdbmsSchemaBuilder,
        entityMetadatas: EntityMetadata[]
      ): Promise<SqlInMemory> {
        const sqlInMemory = new SqlInMemory()

        const args = getPostgresMetadataArgsStorage()
        const triggers = args.triggers
        const triggersToSync: PostgresTrigger[] = []

        triggers?.forEach((trigger) => {
          if (trigger) {
            const target = trigger.target

            const matchingEntity = entityMetadatas.find(
              (entity) => entity.target === target
            )

            if (matchingEntity) {
              triggersToSync.push({
                ...trigger.options,
                tableName: matchingEntity.tableName,
                functionName: this.getFunctionName(trigger.options),
              })
            }
          }
        })

        const existingFunctions = this.existingTriggers.get(schemaBuilder) ?? []

        const functionsQueries = this.getTriggerQueries(queryRunner, {
          existingTriggers: existingFunctions,
          triggersToSync: triggersToSync,
        })

        sqlInMemory.upQueries.push(...functionsQueries.upQueries)
        sqlInMemory.downQueries.push(...functionsQueries.downQueries)

        return Promise.resolve(sqlInMemory)
      }

      private async getCurrentTriggers(
        queryRunner: PostgresQueryRunner
      ): Promise<{name: string; definition: string}[]> {
        const hasTable = await queryRunner.hasTable(this.metadataTable)
        if (!hasTable) {
          return []
        }
        const qb = await queryRunner.connection.createQueryBuilder()

        const metadatas: {name: string; value: string}[] = await qb
          .select()
          .from(this.metadataTable, 't')
          .where(`type = :type`, {
            type: METADATA_TYPE,
          })
          .getRawMany()

        return metadatas.map((metadata) => ({
          name: metadata.name,
          definition: metadata.value,
        }))
      }

      private getTriggerQueries(
        queryRunner: PostgresQueryRunner,
        {
          existingTriggers,
          triggersToSync,
        }: {
          existingTriggers: {name: string; definition: string}[]
          triggersToSync: PostgresTrigger[]
        }
      ): SqlInMemory {
        const sqlInMemory = new SqlInMemory()

        const triggersToDrop: {name: string; definition: string}[] = []
        const triggersToCreate: PostgresTrigger[] = []
        const triggersToUpdate: {
          old: {definition: string}
          new: PostgresTrigger
        }[] = []

        existingTriggers.forEach((existingTrigger) => {
          const triggerToSync = triggersToSync.find(
            (triggerToSync) => existingTrigger.name === triggerToSync.name
          )

          if (!triggerToSync) {
            triggersToDrop.push(existingTrigger)
          } else if (this.hasTriggerChanged(triggerToSync, existingTrigger)) {
            triggersToUpdate.push({
              old: existingTrigger,
              new: triggerToSync,
            })
          }
        })

        triggersToSync.forEach((triggerToSync) => {
          if (
            !existingTriggers.find(
              (trigger) => trigger.name === triggerToSync.name
            )
          ) {
            triggersToCreate.push(triggerToSync)
          }
        })

        triggersToDrop.forEach((triggerToDrop) => {
          sqlInMemory.upQueries.push(this.getDropTriggerQuery(triggerToDrop))
          sqlInMemory.downQueries.push(
            this.getCreateTriggerFromDefinition(triggerToDrop)
          )

          sqlInMemory.upQueries.push(
            queryRunner.deleteTypeormMetadataSql({
              // Typeorm doesn't allow any string in the type parameter - should we reimplement this method?
              type: METADATA_TYPE as any,
              name: triggerToDrop.name,
            })
          )
          sqlInMemory.downQueries.push(
            queryRunner.insertTypeormMetadataSql({
              name: triggerToDrop.name,
              // Typeorm doesn't allow any string in the type parameter - should we reimplement this method?
              type: METADATA_TYPE as any,
              value: triggerToDrop.definition,
            })
          )
        })

        triggersToCreate.forEach((triggerToCreate) => {
          sqlInMemory.upQueries.push(
            this.getCreateTriggerQuery(triggerToCreate)
          )
          sqlInMemory.downQueries.push(
            this.getDropTriggerQuery(triggerToCreate)
          )

          sqlInMemory.upQueries.push(
            queryRunner.insertTypeormMetadataSql({
              name: triggerToCreate.name,
              // Typeorm doesn't allow any string in the type parameter - should we reimplement this method?
              type: METADATA_TYPE as any,
              value: this.getCreateStatement(triggerToCreate),
            })
          )
          sqlInMemory.downQueries.push(
            queryRunner.deleteTypeormMetadataSql({
              // Typeorm doesn't allow any string in the type parameter - should we reimplement this method?
              type: METADATA_TYPE as any,
              name: triggerToCreate.name,
            })
          )
        })

        triggersToUpdate.forEach((trigger) => {
          sqlInMemory.upQueries.push(this.getCreateTriggerQuery(trigger.new))
          sqlInMemory.downQueries.push(
            this.getCreateTriggerFromDefinition(trigger.old)
          )
        })

        return sqlInMemory
      }

      private getCreateTriggerQuery(trigger: PostgresTrigger) {
        return new Query(this.getCreateStatement(trigger))
      }

      private getCreateTriggerFromDefinition(trigger: {definition: string}) {
        return new Query(trigger.definition)
      }

      private getDropTriggerQuery(trigger: {name: string}) {
        return new Query(`DROP TRIGGER IF EXISTS "${trigger.name}"`)
      }

      private getFunctionName(trigger: PostgresTriggerOptions): string {
        if (typeof trigger.execute === 'string') {
          if (trigger.execute.includes('(')) {
            return trigger.execute
          }

          return `${trigger.execute}()`
        }

        const functionTarget = getPostgresMetadataArgsStorage().filterFunctions(
          trigger.execute()
        )

        if (!functionTarget) {
          throw new Error(
            `Cannot find function to execute ${trigger.execute()} for PostgresTrigger ${
              trigger.name
            }`
          )
        }

        return `${functionTarget.options.name}()`
      }

      private hasTriggerChanged(
        trigger: PostgresTrigger,
        oldTrigger: {definition: string}
      ): boolean {
        return this.getCreateStatement(trigger) !== oldTrigger.definition
      }

      private getCreateStatement(trigger: PostgresTrigger): string {
        return `
          CREATE OR REPLACE TRIGGER ${trigger.name}
          ${trigger.when} ${trigger.event.join(' OR ')}
          ON ${trigger.tableName}${
          trigger.referencing
            ? `${
                trigger.referencing.oldTable || trigger.referencing.newTable
                  ? '\nREFERENCING'
                  : ''
              }${
                trigger.referencing.oldTable
                  ? ` OLD TABLE AS ${trigger.referencing.oldTable}`
                  : ''
              }${
                trigger.referencing.newTable
                  ? ` NEW TABLE AS ${trigger.referencing.newTable}`
                  : ''
              }`
            : ''
        }
          ${trigger.forEach === 'row' ? 'FOR EACH ROW' : 'FOR EACH STATEMENT'}${
          trigger.condition ? `\nWHEN (${trigger.condition})` : ''
        }
          EXECUTE FUNCTION ${trigger.functionName};
        `
      }
    }

    return new PostgresExtension()
  }
}
