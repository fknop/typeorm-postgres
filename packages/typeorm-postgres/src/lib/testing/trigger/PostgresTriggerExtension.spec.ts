import {DataSource} from 'typeorm'
import {
  closeTestingDataSource,
  createTestingDataSource,
  initializeDatabaseBeforeAll,
} from '../createTestingDataSource'
import {PostgresTriggerExtension} from '../../extensions/PostgresTriggerExtension'
import {EntityWithTrigger} from './entities/EntityWithTrigger'
import {PostgresFunctionExtension} from '../../extensions/PostgresFunctionExtension'
import {FunctionForTrigger} from './functions/FunctionForTrigger'
import {EntityWithMultipleTriggers} from './entities/EntityWithMultipleTriggers'

describe('PostgresTriggerExtension', () => {
  initializeDatabaseBeforeAll()

  describe('generate migrations for EntityWithTrigger', () => {
    let dataSource: DataSource
    beforeAll(async () => {
      dataSource = await createTestingDataSource({
        entities: [EntityWithTrigger],
        schemaBuilderHooks: [
          PostgresFunctionExtension.init({functions: [FunctionForTrigger]}),
          PostgresTriggerExtension.init(),
        ],
        synchronize: false,
        dropSchema: true,
      })
    })

    afterAll(() => {
      return closeTestingDataSource(dataSource)
    })

    it('should generate proper create migrations', async () => {
      const queries = await dataSource.driver.createSchemaBuilder().log()
      expect(queries).toMatchSnapshot()
    })

    it('should not generate migrations', async () => {
      await dataSource.synchronize(false)
      const queries = await dataSource.driver.createSchemaBuilder().log()

      expect(queries.upQueries).toHaveLength(0)
      expect(queries.downQueries).toHaveLength(0)
    })
  })

  // describe('generate migrations for EntityWithMultipleTriggers', () => {
  //   let dataSource: DataSource
  //   beforeAll(async () => {
  //     dataSource = await createTestingDataSource({
  //       entities: [EntityWithMultipleTriggers],
  //       schemaBuilderHooks: [
  //         PostgresFunctionExtension.init({functions: [FunctionForTrigger]}),
  //         PostgresTriggerExtension.init(),
  //       ],
  //       synchronize: false,
  //       dropSchema: true,
  //     })
  //   })
  //
  //   afterAll(() => {
  //     return closeTestingDataSource(dataSource)
  //   })
  //
  //   it('should generate proper create migrations', async () => {
  //     const queries = await dataSource.driver.createSchemaBuilder().log()
  //     expect(queries).toMatchSnapshot()
  //   })
  //
  //   it('should not generate migrations', async () => {
  //     await dataSource.synchronize(false)
  //     const queries = await dataSource.driver.createSchemaBuilder().log()
  //
  //     expect(queries.upQueries).toHaveLength(0)
  //     expect(queries.downQueries).toHaveLength(0)
  //   })
  // })
})
