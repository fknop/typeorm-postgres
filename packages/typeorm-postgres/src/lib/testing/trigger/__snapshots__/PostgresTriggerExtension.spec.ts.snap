// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostgresTriggerExtension generate migrations for EntityWithTrigger should generate proper create migrations 1`] = `
SqlInMemory {
  "downQueries": Array [
    Query {
      "@instanceof": Symbol(Query),
      "parameters": undefined,
      "query": "DROP FUNCTION IF EXISTS \\"prevent_update\\"",
    },
    Query {
      "@instanceof": Symbol(Query),
      "parameters": Array [
        "POSTGRES_FUNCTION",
        "prevent_update",
      ],
      "query": "DELETE FROM \\"public\\".\\"typeorm_metadata\\" WHERE \\"type\\" = $1 AND \\"name\\" = $2",
    },
    Query {
      "@instanceof": Symbol(Query),
      "parameters": undefined,
      "query": "DROP TABLE \\"entity_with_trigger\\"",
    },
    Query {
      "@instanceof": Symbol(Query),
      "parameters": undefined,
      "query": "DROP TRIGGER IF EXISTS \\"prevent_update_entity\\"",
    },
    Query {
      "@instanceof": Symbol(Query),
      "parameters": Array [
        "POSTGRES_TRIGGER",
        "prevent_update_entity",
      ],
      "query": "DELETE FROM \\"public\\".\\"typeorm_metadata\\" WHERE \\"type\\" = $1 AND \\"name\\" = $2",
    },
  ],
  "upQueries": Array [
    Query {
      "@instanceof": Symbol(Query),
      "parameters": undefined,
      "query": "
    CREATE OR REPLACE FUNCTION prevent_update() RETURNS trigger AS $$
    BEGIN
      IF TG_OP = 'UPDATE' THEN
        RAISE 'cannot update this record';
      END IF;
    END;
    $$ LANGUAGE plpgsql;
  ",
    },
    Query {
      "@instanceof": Symbol(Query),
      "parameters": Array [
        "POSTGRES_FUNCTION",
        "prevent_update",
        "
    CREATE OR REPLACE FUNCTION prevent_update() RETURNS trigger AS $$
    BEGIN
      IF TG_OP = 'UPDATE' THEN
        RAISE 'cannot update this record';
      END IF;
    END;
    $$ LANGUAGE plpgsql;
  ",
      ],
      "query": "INSERT INTO \\"public\\".\\"typeorm_metadata\\"(\\"database\\", \\"schema\\", \\"table\\", \\"type\\", \\"name\\", \\"value\\") VALUES (DEFAULT, DEFAULT, DEFAULT, $1, $2, $3)",
    },
    Query {
      "@instanceof": Symbol(Query),
      "parameters": undefined,
      "query": "CREATE TABLE \\"entity_with_trigger\\" (\\"id\\" uuid NOT NULL DEFAULT uuid_generate_v4(), CONSTRAINT \\"PK_58d4ce0f1dc13c9591eb032db7b\\" PRIMARY KEY (\\"id\\"))",
    },
    Query {
      "@instanceof": Symbol(Query),
      "parameters": undefined,
      "query": "
          CREATE OR REPLACE TRIGGER prevent_update_entity
          BEFORE UPDATE
          ON entity_with_trigger
          FOR EACH STATEMENT
          EXECUTE FUNCTION prevent_update();
        ",
    },
    Query {
      "@instanceof": Symbol(Query),
      "parameters": Array [
        "POSTGRES_TRIGGER",
        "prevent_update_entity",
        "
          CREATE OR REPLACE TRIGGER prevent_update_entity
          BEFORE UPDATE
          ON entity_with_trigger
          FOR EACH STATEMENT
          EXECUTE FUNCTION prevent_update();
        ",
      ],
      "query": "INSERT INTO \\"public\\".\\"typeorm_metadata\\"(\\"database\\", \\"schema\\", \\"table\\", \\"type\\", \\"name\\", \\"value\\") VALUES (DEFAULT, DEFAULT, DEFAULT, $1, $2, $3)",
    },
  ],
}
`;
